internal/store/vpa.go

```diff
@@ -0,0 +1,421 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package store
+
+import (
+	"fmt"
+
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/watch"
+
+	autoscaling "k8s.io/autoscaler/vertical-pod-autoscaler/pkg/apis/autoscaling.k8s.io/v1beta2"
+	vpaclientset "k8s.io/autoscaler/vertical-pod-autoscaler/pkg/client/clientset/versioned"
+	clientset "k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/kube-state-metrics/pkg/metric"
+)
+
+var (
+	descVerticalPodAutoscalerLabelsName          = "kube_vpa_labels"
+	descVerticalPodAutoscalerLabelsHelp          = "Kubernetes labels converted to Prometheus labels."
+	descVerticalPodAutoscalerLabelsDefaultLabels = []string{"namespace", "vpa", "target_ref"}
+
+	vpaMetricFamilies = []metric.FamilyGenerator{
+		{
+			Name: descVerticalPodAutoscalerLabelsName,
+			Type: metric.Gauge,
+			Help: descVerticalPodAutoscalerLabelsHelp,
+			GenerateFunc: wrapVPAFunc(func(a *autoscaling.VerticalPodAutoscaler) *metric.Family {
+				labelKeys, labelValues := kubeLabelsToPrometheusLabels(a.Labels)
+				return &metric.Family{
+					Metrics: []*metric.Metric{
+						{
+							LabelKeys:   labelKeys,
+							LabelValues: labelValues,
+							Value:       1,
+						},
+					},
+				}
+			}),
+		},
+		{
+			Name: "kube_vpa_update_mode",
+			Type: metric.Gauge,
+			Help: "Update mode of the VPA.",
+			GenerateFunc: wrapVPAFunc(func(a *autoscaling.VerticalPodAutoscaler) *metric.Family {
+				ms := []*metric.Metric{}
+
+				if a.Spec.UpdatePolicy.UpdateMode == nil {
+					return &metric.Family{
+						Metrics: ms,
+					}
+				}
+
+				for _, mode := range []autoscaling.UpdateMode{
+					autoscaling.UpdateModeOff,
+					autoscaling.UpdateModeInitial,
+					autoscaling.UpdateModeRecreate,
+					autoscaling.UpdateModeAuto,
+				} {
+					var v float64
+					if *a.Spec.UpdatePolicy.UpdateMode == mode {
+						v = 1
+					} else {
+						v = 0
+					}
+					ms = append(ms, &metric.Metric{
+						LabelKeys:   []string{"update_mode"},
+						LabelValues: []string{string(mode)},
+						Value:       v,
+					})
+				}
+
+				return &metric.Family{
+					Metrics: ms,
+				}
+			}),
+		},
+		{
+			Name: "kube_vpa_container_resource_policy_min_cpu_cores",
+			Type: metric.Gauge,
+			Help: "Minimum CPU cores the VPA can set for containers matching the name.",
+			GenerateFunc: wrapVPAFunc(func(a *autoscaling.VerticalPodAutoscaler) *metric.Family {
+				ms := []*metric.Metric{}
+				for _, c := range a.Spec.ResourcePolicy.ContainerPolicies {
+					min := c.MinAllowed
+					if cpu, ok := min[v1.ResourceCPU]; ok {
+						ms = append(ms, &metric.Metric{
+							LabelKeys:   []string{"container_name"},
+							LabelValues: []string{c.ContainerName},
+							Value:       float64(cpu.MilliValue()) / 1000,
```

Comment:
I believe the resource type can do the conversion internally, I don't think we have to do the math here

Reply:
If I take the float64 conversion off I get:
```
cannot use val.MilliValue() / 1000 (type int64) as type float64 in field value
```